### PR TITLE
Dispose response messages when we retry sending requests. 

### PR DESCRIPTION
**Pull request by TianoMS**
Fix issue: https://github.com/Azure/azure-powershell/issues/1961

### CODE REVIEW

**Comment by TianoMS** [comment]
On demand job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/886/

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs [hunk 1]
```diff
@@ -56,15 +56,15 @@ public class RetryHandler : DelegatingHandler
         protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
             HttpResponseMessage response = null;
+
             for (int attempt = 0; attempt < RetryHandler.MaxAttempts; ++attempt)
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: This is not very readable, IMHO. We should just do this:
for(int attempt = 1; attempt <= RetryHandler.MaxAttempts; ++attempt)
{
..
}

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs [hunk 2]
```diff
@@ -56,15 +56,15 @@ public class RetryHandler : DelegatingHandler
...
                 try
                 {
-                    response = await base
-                        .SendAsync(request: request, cancellationToken: cancellationToken)
+                    response = await base.SendAsync(request: request, cancellationToken: cancellationToken)
                         .ConfigureAwait(continueOnCapturedContext: false);
-
-                    if (attempt == RetryHandler.MaxAttempts ||
+                    
+                    if (attempt == RetryHandler.MaxAttempts - 1 ||
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Why do we need the attempt == RetryHandler.MaxAttempts -1 condition here? The loop condition will ensure attempt won't go above MaxAttempts (3).

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs [hunk 3]
```diff
@@ -74,17 +74,22 @@ public class RetryHandler : DelegatingHandler
                 }
                 catch (Exception ex)
                 {
-                    if (ex.IsFatal() || attempt == RetryHandler.MaxAttempts)
+                    if (ex.IsFatal() || attempt == RetryHandler.MaxAttempts - 1)
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Please see my comment above, regarding changing the for loop condition. If we keep it as for(int attempt = 1; attempt <= RetryHandler.MaxAttempts; ++attempt), this should just read attempt == RetryHandler.MaxAttempts

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs [hunk 4]
```diff
@@ -74,17 +74,22 @@ public class RetryHandler : DelegatingHandler
...
+                {
+                    response.Dispose();
+                }
+
                 await Task.Delay(delay: RetryHandler.GetDelay(attempt), cancellationToken: cancellationToken)
                     .ConfigureAwait(continueOnCapturedContext: false);
             }
 
-            return response;
+            return null;
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Why? Why not response?

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs [hunk 5]
```diff
@@ -219,27 +219,39 @@ public ResourceManagerRestClientBase(Components.HttpClientHelper httpClientHelpe
...
+                            errorMessage: message);
+                    }
+
+                    return response;
                 }
+                catch (Exception exception)
+                {
+                    if (exception is OperationCanceledException && !cancellationToken.IsCancellationRequested)
+                    {
+                        throw new TimeoutException("A request times out.");
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs]: Can we throw a more detailed exception, instead of just "A request timed out" ? What was the request? What is the content of exception.Message?

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 2 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Previously the condition was "attempt == RetryHandler.MaxAttempts", which would never be true. I think the point here is directly return the response after MaxAttempts of tries, and don't do the rest of the stuff.

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 4 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: We should actually never reach here. The response would already have been returned from the loop after MaxAttempts tries.

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs (hunk 5 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs]: When we get here, the inner exception is "A task was canceled". Here we are trying to catch on "await base.SendAsync(...).ConfigureAwait(...)".

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 4 above)
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: The return is needed for this to successfully compile. I'd keep this as response (which is set to null either way), instead of returning null explicitly

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 2 above)
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Per your logic, we will return response in the 3rd attempt, which is not what we want.

Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Hmm... I thought we want to attempt three times in total..

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 1 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Fixed.

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 3 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: This is fixed.

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 4 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: Added a comment in code.

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 3 above)
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: How can attempt > RetryHandler.MaxAttempts in for (int attempt = 1; attempt <= RetryHandler.MaxAttempts; ++attempt) ? :)  

---
In reply to: [66539492](https://github.com/Azure/azure-powershell/pull/2412#discussion_r66539492) [](ancestors = 66539492)

**Comment by gandhiniraj** [comment]
>                         return response;

For better readability and not explicitly return null value, how about this ?
- do break; here
- instead of return null just do return response;


---
Refers to: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs:72 in 73325f4. [](commit_id = 73325f4f64d08c6ab32b7fc3689052d05b7fc782, deletion_comment = False)

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 3 above)
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: And combining with my response above if we want to throw if we have reached max attempt, initialize attempt outside of the for loop and if  attempt > RetryHandler.MaxAttempts then throw outside of the for loop

---
In reply to: [66560166](https://github.com/Azure/azure-powershell/pull/2412#discussion_r66560166) [](ancestors = 66560166,66539492)

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs [hunk 6]
```diff
@@ -219,27 +220,39 @@ public ResourceManagerRestClientBase(Components.HttpClientHelper httpClientHelpe
...
-                    .SendAsync(request: request, cancellationToken: cancellationToken)
-                    .ConfigureAwait(continueOnCapturedContext: false);
-
-                if (!response.StatusCode.IsSuccessfulRequest())
+                try
                 {
-                    var errorResponse = await ResourceManagerRestClientBase
-                        .TryReadErrorResponseMessage(response, rewindContentStream: true)
+                    var response = await httpClient
+                        .SendAsync(request: request, cancellationToken: cancellationToken)
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs]: nit:For better readability use: await httpClient.SendAsync(request: request, cancellationToken: cancellationToken) and next line .Configure...

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs [hunk 7]
```diff
@@ -219,27 +220,39 @@ public ResourceManagerRestClientBase(Components.HttpClientHelper httpClientHelpe
...
                         .ConfigureAwait(continueOnCapturedContext: false);
 
-                    var message = await ResourceManagerRestClientBase
-                        .GetErrorMessage(request: request, response: response, errorResponse: errorResponse)
-                        .ConfigureAwait(continueOnCapturedContext: false);
+                    if (!response.StatusCode.IsSuccessfulRequest())
+                    {
+                        var errorResponse = await ResourceManagerRestClientBase
+                            .TryReadErrorResponseMessage(response, rewindContentStream: true)
+                            .ConfigureAwait(continueOnCapturedContext: false);
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs]: nit: same as above

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.resx [hunk 8]
```diff
@@ -396,4 +396,7 @@
   <data name="UpdatedResourceGroup" xml:space="preserve">
     <value>Updated resource group '{0}' in location '{1}'</value>
   </data>
+  <data name="OperationFailedWithTimeOut" xml:space="preserve">
+    <value>Operation failed because a request time out.</value>
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.resx]: nit: the request timed out.

Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.resx]: Checked wiki. It's a verb. "timed out" is correct.

Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.resx]: I am not saying it's not a word - I am just saying grammatically - "Operation failed because the request timed out"

---
In reply to: [66566451](https://github.com/Azure/azure-powershell/pull/2412#discussion_r66566451) [](ancestors = 66566451)

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs (hunk 3 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs]: attempt shouldn't be > MaxAttempts. my latest change uses attempt == MaxAttempts as exit criteria.

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.resx (hunk 8 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.resx]: Sorry for the confusion... I was actually agreeing with you :)  I'm changing it to "timed out".

Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.resx]: There isn't any confusion. all good

---
In reply to: [66567017](https://github.com/Azure/azure-powershell/pull/2412#discussion_r66567017) [](ancestors = 66567017)

**Comment by TianoMS** [comment]
>                         return response;

Will update it. 

---
In reply to: [225104925](https://github.com/Azure/azure-powershell/pull/2412#issuecomment-225104925) [](ancestors = 225104925,225090348)

---
Refers to: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Handlers/RetryHandler.cs:72 in 73325f4. [](commit_id = 73325f4f64d08c6ab32b7fc3689052d05b7fc782, deletion_comment = False)

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs (hunk 6 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs]: Checked server code. Our practice seems to be putting them at separate lines. Will follow the practice here.

**Comment by TianoMS** [comment]
On demand job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/893/

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.Designer.cs [hunk 9]
```diff
@@ -493,6 +493,15 @@ internal class Resources {
         }
         
         /// <summary>
+        ///   Looks up a localized string similar to Operation failed because a request timed out..
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.Designer.cs]: nit: operation

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs (hunk 6 above)
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs]: sounds good

---
In reply to: [66568621](https://github.com/Azure/azure-powershell/pull/2412#discussion_r66568621) [](ancestors = 66568621)

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.Designer.cs (hunk 9 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Properties/Resources.Designer.cs]: This is generated code. Looks it uses the original resource string.

**Comment by gandhiniraj** [comment]
>         protected async Task<HttpResponseMessage> SendRequestAsync(HttpRequestMessage request,

we should add a test for this

---
Refers to: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs:218 in c591214. [](commit_id = c59121483281fa3d350ea08f380c9671fadd66aa, deletion_comment = False)

**Comment by TianoMS** [comment]
>         protected async Task<HttpResponseMessage> SendRequestAsync(HttpRequestMessage request,

This is a corner case. It happens when an async task is stuck and eventually times out. When this happens we catch the exception and throw a more readable exception. I haven't figured out a good way to test this, but I'll try this later as separate efforts. 

---
In reply to: [226050060](https://github.com/Azure/azure-powershell/pull/2412#issuecomment-226050060) [](ancestors = 226050060)

---
Refers to: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/RestClients/ResourceManagerRestClientBase.cs:218 in c591214. [](commit_id = c59121483281fa3d350ea08f380c9671fadd66aa, deletion_comment = False)

**Comment by TianoMS** [comment]
@markcowl All comments were addressed except one on test coverage. I didn't figure out a good way to test the specific piece of fix yet. Regression is covered, though. I think I'll try to figure that out as separate tasks.
